From 589cb7039d7774f13781a8ca4783fb994c5ee096 Mon Sep 17 00:00:00 2001
From: Hari Chandrakanthan <quic_haric@quicinc.com>
Date: Mon, 7 Nov 2022 11:33:50 +0530
Subject: [PATCH] ath12k : do not drop data frames from unassociated stations
 in the driver

When data frames from unassociated stations are received by an AP,
the AP is supposed to send a deauth/disassoc frame with reason code
"Class 2 frame received from nonauthenticated STA" or
"Class 3 frame received from nonassociated STA".

Since the ath12k driver drops the data frames from unassociated
station, the upper layer(mac80211/hostapd) is not aware of such event.
Hence deauth/disassoc frame is not sent to that particular station.

Allow the data frames from unassociated stations to reach mac80211
so that mac80211 can send NL80211_CMD_UNEXPECTED_FRAME
event to userspace(hostapd) and the data frame shall be dropped
by mac80211.

hostapd upon receiving the event NL80211_CMD_UNEXPECTED_FRAME
will be sending the disassoc frame with proper reason code.

Signed-off-by: Hari Chandrakanthan <quic_haric@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 4074633..74d2128 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3987,12 +3987,6 @@ static int ath12k_dp_rx_h_null_q_desc(struct ath12k *ar, struct sk_buff *msdu,
 	msdu_len = ath12k_dp_rx_h_msdu_len(ab, desc);
 	peer_id = ath12k_dp_rx_h_peer_id(ab, desc);
 
-	if (!ath12k_peer_find_by_id(ab, peer_id)) {
-		ath12k_dbg(ab, ATH12K_DBG_DATA, "invalid peer id received in wbm err pkt%d\n",
-			   peer_id);
-		return -EINVAL;
-	}
-
 	if (!rxcb->is_frag && ((msdu_len + hal_rx_desc_sz) > DP_RX_BUFFER_SIZE)) {
 		/* First buffer will be freed by the caller, so deduct it's length */
 		msdu_len = msdu_len - (DP_RX_BUFFER_SIZE - hal_rx_desc_sz);
-- 
2.7.4

